sw/otbn/crypto/p256_mod_inv_sca.s

```diff
@@ -0,0 +1,141 @@
+/* Copyright lowRISC contributors. */
+/* Licensed under the Apache License, Version 2.0, see LICENSE for details. */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+/**
+ * Wrapper specifically for SCA/formal analysis of p256 modular inversion.
+ *
+ * This routine would never normally be exposed, but it's helpful for SCA to
+ * analyze it in isolation. 
+ */
+
+.section .text.start
+
+main:
+  /* Initialize all-zero register. */
+  bn.xor    w31, w31, w31
+
+  /* Load modulus and Barrett constant.
+       w28 <= u, Barrett constant for n
+       w29 <= n, p256 curve order
+       MOD <= n */
+  li        x2, 28
+  la        x3, p256_u_n
+  bn.lid    x2, 0(x3)
+  li        x2, 29
+  la        x3, p256_n
+  bn.lid    x2, 0(x3)
+  bn.wsrw   0, w29
+
+  /* Load first share of input.
+       w0, w1 <= dmem[k0] */
+  la        x16, k0
+  li        x2, 0
+  bn.lid    x2, 0(x16++)
+  li        x2, 1
+  bn.lid    x2, 0(x16)
+
+  /* Load second share of input.
+       w2, w3 <= dmem[k1] */
+  la        x16, k1
+  li        x2, 2
+  bn.lid    x2, 0(x16++)
+  li        x2, 3
+  bn.lid    x2, 0(x16)
+
+  /* NOTE: the masking logic is written to mimic the version in `p256_sign`
+     from p256.s; see the documentation of that function for details. */
+
+  /* Generate a random 127-bit number.
+       w4 <= URND()[255:129] */
+  bn.wsrr  w4, 0x2 /* URND */
+  bn.rshi  w4, w31, w4 >> 129
+
+  /* Add 1 to get a 128-bit nonzero scalar for masking.
+       w4 <= w4 + 1 = alpha */
+  bn.addi  w4, w4, 1
```

Comment:
Is this a good way to generate a 128-bit nonzero random value? After above lines we will have a 128-bit mask with its MSB 0 with a very high probability, right? Is this something acceptable for the rest of the algorithm?

Reply:
Yes, this is effectively only 127 bits of masking. But 127 bits of (multiplicative) masking should be plenty -- my estimate here is that 64 bits would probably be enough, but on OTBN it's pretty cheap to get more bits so we might as well go all the way to 127.